lib/collection/src/shard/mod.rs

```diff
@@ -0,0 +1,71 @@
+pub mod local_shard;
+pub mod remote_shard;
+
+use crate::collection_manager::holders::segment_holder::SegmentHolder;
+use crate::shard::remote_shard::RemoteShard;
+use crate::{
+    CollectionInfo, CollectionResult, CollectionSearcher, CollectionUpdateOperations, LocalShard,
+    OptimizersConfigDiff, Record, UpdateResult,
+};
+use async_trait::async_trait;
+use parking_lot::RwLock;
+use segment::types::{ExtendedPointId, Filter, WithPayloadInterface};
+use std::sync::Arc;
+
+pub type ShardId = u32;
+
+pub type PeerId = u32;
+
+/// Shard
+///
+/// A shard can either be local or remote
+///
+#[allow(clippy::large_enum_variant)]
+pub enum Shard {
+    Local(LocalShard),
+    Remote(RemoteShard),
+}
+
+impl Shard {
+    pub fn get(&self) -> Arc<dyn ShardOperation + Sync + Send + '_> {
+        match self {
+            Shard::Local(local_shard) => Arc::new(local_shard),
+            Shard::Remote(remote_shard) => Arc::new(remote_shard),
+        }
+    }
+
+    pub async fn before_drop(&mut self) {
```

Comment:
`before_drop` cannot be part of the `ShardOperation` trait due to the `&mut`.
As it is kind of a snowflake operation, I decided to implement the polymorphism it at the level of the enum.

Reply:
I don't think we will need to manually drop anything for the Remote shard, so this should be ok I think